### PR TITLE
[LOTUS-5792] Error inserting nonnullable

### PR DESCRIPTION
**Pull request by jdavenport-sm**
Certain very specific sequences of changefeed messages could result in data from one record being inappropriately combined with data from another record, possibly even from a different table. This would result in SQL errors and EDS exit.

This was caused by an optimization that would attempt to combine inserts, updates, and/or deletes on the same record. In some cases, a delete would simply remove a prior update from the list of records in a batch. The problem stemmed from a parallel data structure that tracked the primary keys of records in the batch, and, importantly, the index in the records array that those primary keys represented. When a delete removed a record from the batch, the indices in this parallel data structure were not updated. This could cause unrelated records to be inappropriately batched which would result in an SQL error when the Snowflake driver attempted to insert incompatible data into a table column. The observed result was a failure on attempting to insert null into a non-nullable field.

I have created a test that catches this problem. Refer to this test for more details, as the sequence of events required to cause the error is too nuanced to easily describe in this message. One simple solution is to clear the PKs->record indices map (the parallel data structure from above) when a delete is batched. If possible I intend to separate the optimization from the batcher add function so the code is better encapsulated. Another possible improvement will be to move some or all of the optimizations to a separate step and location to improve reliability of EDS.

Just because this is cumbersome to actually test, I made a video of me walking through how it works:
https://www.loom.com/edit/115eec05fdb24db6a5b9c582f2c61d43

---

> [!NOTE]
> Replaces in-batcher record combining with a new safe bulk optimizer applied during Snowflake flush, adds tests, simplifies batcher, and updates release signing key.
> 
> - **Snowflake Driver**:
>   - Apply bulk record modifications before SQL generation via `runBulkRecordModifications`, initialized with `util.CombineRecordsWithSamePrimaryKey`.
> - **Batching/Optimization**:
>   - Simplify `util.Batcher.Add` to append-only; remove PK index tracking and in-place combining logic.
>   - Introduce `util.CombineRecordsWithSamePrimaryKey` to safely collapse multiple updates/deletes for the same PK without cross-record data leakage.
> - **Tests**:
>   - Add `internal/util/optimize_test.go` covering update merge, delete handling, and edge cases.
>   - Update `internal/util/batcher_test.go` to reflect append-only behavior.
>   - Minor variable updates in `snowflake/sql_test.go`.
> - **Build/Release**:
>   - Update `.goreleaser.yaml` quoting and change signing key to `3C246E2D0C4F9134`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a1eefb0cce58d69903b3f113be6a5ebb4784a3ec. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: internal/util/optimize.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+					if !slices.Contains(previousRecord.Diff, key) {
+						previousRecord.Diff = append(previousRecord.Diff, key)
+					}
+				}
+				previousRecord.Event = record.Event
+				maps.Copy(previousRecord.Object, record.Object)
+			case deleteWithBatch:
+				collapsedRecords = collapsedRecords[:0]
+				collapsedRecords = append(collapsedRecords, record)
+				previousRecord = record
```
[the remark this answers is not in the document]
Comment (jdavenport-sm) [reply internal/util/optimize.go:57]: This is the desired behavior: the records before the delete are dropped, the delete is kept, then in the next iteration of the loop, the delete is the previous statement. Because we are iterating over records, not collapsedRecords, the modification to the slice isn't a problem.
